File: data-structures/LinkedList/python/LinkedList.py
class Node(object):
    def __init__(self, data):
        self.data = data
        self.next = None

class LinkedList(object):
    def __init__(self):
        self.head = None

    def prepend(self,data):
        node = Node(data)
        if self.head:
            node.next = self.head
        self.head = node

    def append(self,data):
        node = Node(data)
        if not self.head:
            self.head = node
            return
        last = self.head
        while last.next:
            last = last.next
        last.next = node
        
    def search(self,data):
        node = self.head
        while node:
            if node.data == data:
                return True;
            node = node.next
        return False

    def remove(self,data):
        if not self.head:
            return False
        if self.head.data == data:
            self.head = self.head.next
        node = self.head
        while node.next:
            if node.next.data == data:
                node.next = node.next.next
                return True
            node = node.next
        return False

    def first(self):
        return self.head

    def last(self):
        if not self.head:
            return None
        node = self.head
        while node.next:
            node = node.next
        return node

    def size(self):
        count = 0
        node = self.head
        while node:
            count += 1
        return count
        
    def display(self):
        node = self.head
        while node:
            print(node.data)
            node = node.next

    def toArray(self):
        node = self.head
        l = []
        while node:
            l.append(node.data)
            node = node.next
        print(l)

    def reverse(self):
        cur = self.last()
        while cur != self.head:
            prev = self.head
            while prev.next != cur:
                prev = prev.next
            print(cur.data)
            cur = prev
        print(cur.data)

if __name__=='__main__':
    list = LinkedList()
    list.append(2)
    list.append(4)
    list.append(6)
    list.append(7)
    list.append(9)
    list.append(10)

    #list.display()
    #print(list.first().data,list.last().data)
    #list.toArray()
    list.remove(9)
    list.display()
    print("\n")
    list.reverse()

File: data-structures/LinkedList/go/LinkedList.go
package main

import (
	"fmt"
)

type Node struct {
	data interface{}
	next *Node
}

type LinkedList struct {
	head *Node
}

func (list *LinkedList) prepend(data interface{}) {
	node := &Node{data: data, next: nil}
	if list.head != nil {
		node.next = list.head
	}
	list.head = node
}

func (list *LinkedList) append(data interface{}) {
	node := &Node{data: data, next: nil}
	if list.head == nil {
		list.head = node
		return
	}
	cur := list.head
	for cur.next != nil {
		cur = cur.next
	}
	cur.next = node
}

func (list *LinkedList) search(data interface{}) bool {
	if list.head == nil {
		return false
	}
	node := list.head
	for node != nil {
		if node.data == data {
			return true
		}
		node = node.next
	}
	return false
}

func (list *LinkedList) remove(data interface{}) bool {
	if list.head == nil {
		return false
	}
	if list.head.data == data {
		list.head = list.head.next
		return true
	}
	node := list.head
	for node.next != nil {
		if node.next.data == data {
			node.next = node.next.next
			return true
		}
		node = node.next
	}
	return false
}

func (list *LinkedList) size() int {
	count := 0
	node := list.head
	for node != nil {
		node = node.next
		count++
	}
	return count
}

func (list *LinkedList) first() *Node {
	return list.head
}

func (list *LinkedList) last() *Node {
	node := list.head
	for node.next != nil {
		node = node.next
	}
	return node
}

func (list *LinkedList) display() {
	node := list.head
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

func (list *LinkedList) reverse() {
	cur := list.last()
	for cur != list.head {
		prev := list.head
		for prev.next != cur {
			prev = prev.next
		}
		fmt.Println(cur.data)
		cur = prev
	}
	fmt.Println(cur.data)
}

func main() {
	list := LinkedList{}
	list.append(2)
	list.append(4)
	list.append(6)
	list.append(7)
	list.append(9)
	list.append(10)

	//list.display()
	fmt.Println(list.first(), list.last(), list.size(), list.search(19), list.remove(14))
	fmt.Println()
	list.reverse()
	//list.display()
}
